### PR TITLE
Problem: wrong errno set when feature is missing

### PR DESCRIPTION
**Pull request by paddor**
Solution: set errno to ENOTSUP

### CODE REVIEW

**Comment by bluca** [comment]
Thanks for the pr but unfortunately these are API breakages so they can only be done for draft ones - could you please rebase and take out the changes to the stable APIs? Thanks

**Comment by paddor** [comment]
DISH/SERVER/... are stable now? Or which ones do you mean?

As for the cases where `NULL` is returned, any `errno` value read by client code was a previously set one.

**Comment by paddor** [comment]
I briefly stumbled across this problem in #1807 (see bottom of issue description). And today I wasted time because my tests got an EINVAL when setting the group on a frame until I realized that CMZQ got installed without DRAFT APIs again (this time with brew on Mac, which is another problem). 

Sorry but EINVAL just doesn’t make sense in these cases and not setting an errno at all (in the constructors) is no help when debugging.

**Comment by bluca** [comment]
> I just checked the API files. Seems like SERVER/CLIENT/RADIO/DISH are still draft.

Yes those changes are all fine - but xpub/xsub/stream are stable so we cannot change the errno. I know enotsup makes way more sense but it's too late I'm afraid.

File: src/zsock.c [hunk 1]
```diff
@@ -2165,6 +2174,16 @@ zsock_test (bool verbose)
     zsock_destroy (&client);
     zsock_destroy (&server);
 
+#else
+    errno = 0;
+    zsock_t* server = zsock_new_server (NULL);
```
Comment (bluca) [inline src/zsock.c]:
```
src/zsock.c: In function ‘zsock_test’:
src/zsock.c:2179:14: error: redefinition of ‘server’
     zsock_t* server = zsock_new_server (NULL);
              ^
src/zsock.c:1924:14: note: previous definition of ‘server’ was here
     zsock_t *server = zsock_new (ZMQ_DEALER);
```

This breaks the tests, could you please fix it? Thanks!

**Comment by paddor** [comment]
Thanks for the clarification. Updated.

I'm curious though: How is an undefined value in `errno` part of the stable API? To my understanding, it used to be whatever it was set to in a previous call. Are you thinking of programs depending on that undefined value, as long as it is deterministic?

File: src/zsock.c [hunk 2]
```diff
@@ -321,6 +322,7 @@ zsock_new_xsub_checked (const char *endpoints, const char *filename, size_t line
         zsock_destroy (&sock);
     return sock;
 #else
+    errno = ENOTSUP;
```
Comment (bluca) [inline src/zsock.c]: This one too please :-)

File: src/zsock.c [hunk 3]
```diff
@@ -363,6 +365,7 @@ zsock_new_stream_checked (const char *endpoints, const char *filename, size_t li
         zsock_destroy (&sock);
     return sock;
 #else
+    // errno = ENOTSUP;      // too late, stable API would be broken
```
Comment (bluca) [inline src/zsock.c]: Ditto

File: src/zsock.c (hunk 2 above)
Comment (paddor) [reply src/zsock.c]: Sorry, missed this one. Updated.

**Comment by bluca** [comment]
> I'm curious though: How is an undefined value in errno part of the stable API? To my understanding, it used to be whatever it was set to in a previous call. Are you thinking of programs depending on that undefined value, as long as it is deterministic?

Going from "does not set an error" to "sets this error in this case" is an interface change that would cause users of that api to have to at the very least re evaluate their usage when they read the changelog, which is something that a stable library should not cause

File: src/zsock.c [hunk 4]
```diff
@@ -1912,17 +1921,17 @@ zsock_test (bool verbose)
     assert (port >= 60000 && port <= 60500);
 
     //  Test zsock_attach method
-    zsock_t *server = zsock_new (ZMQ_DEALER);
```
Comment (bluca) [inline src/zsock.c]: Sorry to be a pain in the back but CI compilation is still broken:

```
src/zsock.c: In function 'zsock_test':
src/zsock.c:2111:5: error: 'server' undeclared (first use in this function)
     server = zsock_new (ZMQ_SERVER);
```

There's more usage of that variable below, so perhaps it's easier to use different names in the new snippet? Up to you though

Comment (paddor) [reply src/zsock.c]: No problem. Fixed and running `make && make check` this time. Stay tuned...

Comment (paddor) [reply src/zsock.c]: Okay, this time it should be good. Sorry about this, I'm too used to Ruby. 😆

**Comment by paddor** [comment]
Thanks!

**Comment by bluca** [comment]
Thanks to you for the fixes and being patient :)

**Pull request merged**